packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
+                }
+            }
+            return;
+        }
 
+         // Handle liquidity tokens
         for (uint i = 0; i < ctx.size; i++) {
-            uint96 amount = amounts[i];
+            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].add(amounts[i]);
+
+            if (joinFromLayer1) {
+                // Do nothing
+            } else {
+                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].add(amounts[i]);
 
-            if (join.fromLayer2) {
                 TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
                 ctx.numTransactionsConsumed++;
 
                 // We do not check these fields: fromAccountID, to, amount, fee, storageID
                 require(
-                    transfer.toAccountID == S.accountID &&
+                    transfer.toAccountID== ctx.accountID &&
                     transfer.from == join.owner &&
                     transfer.tokenID == ctx.tokens[i].tokenID &&
-                    transfer.amount.isAlmostEqual(amount) &&
+                    transfer.amount.isAlmostEqual(amounts[i]) &&
                     transfer.feeTokenID == ctx.tokens[i].tokenID &&
-                    transfer.fee.isAlmostEqual(join.fees[i]),
+                    transfer.fee.isAlmostEqual(join.joinFees[i]),
                     "INVALID_TX_DATA"
                 );
 
-                // Replay protection when using a signature (otherwise the approved hash is cleared onchain)
-                if (signature.length > 0) {
-                    require(transfer.storageID == join.storageIDs[i], "INVALID_TX_DATA");
+                if (i == 0 && signature.length > 0) {
+                    require(transfer.storageID == join.joinStorageID, "INVALID_STORAGE_ID_OR_REPLAY");
                 }
 
-                // Now approve this transfer
                 transfer.validUntil = 0xffffffff;
                 bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
                 ctx.exchange.approveTransaction(join.owner, txHash);
+            }
+        }
 
-                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].add(amount);
+        // Handle pool token
+        ctx.totalSupply = ctx.totalSupply.add(mintAmount);
 
-            } else {
-                // Make the amount unavailable for withdrawing
-                address token = ctx.tokens[i].addr;
-                S.userBalance[token][join.owner] = S.userBalance[token][join.owner].sub(amount);
-            }
+        bool mintToLayer1 = (
+            join.direction == AmmData.Direction.L1_TO_L1 ||
+            join.direction == AmmData.Direction.L2_TO_L1);
 
-            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].add(amount);
+        if (mintToLayer1) {
+            S.mint(join.owner, mintAmount);
+        } else {
+            _approvePoolTokenDeposit(ctx, mintAmount, join.owner);
```

Comment:
Seems like we should still mint the tokens here to `address(this)` so they can be deposited to the user.